download.sh

```diff
@@ -0,0 +1,101 @@
+
+Arguments:
+    <DUMP_DIR>  An existing directory to store dumps in. Dumps will be grouped
+                into subdirectories by date, and a link 'latest' will point to
+                the latest complete dump subdirectory, if it exists.
+
+Environment Variables:
+    LANGUAGES   A whitespace-separated list of wikipedia language codes to
+                download dumps of.
+                Defaults to the languages in 'article_processing_config.json'.
+                See <https://meta.wikimedia.org/wiki/List_of_Wikipedias>.
+
+Exit codes:
+    0   The lastest dumps are already present or were downloaded successfully.
+    1   Argument error.
+    16  Some of languages were not available to download. The latest dump may
+        be in progress, or some of the specified languages may not exist.
+    _   Subprocess error.
+"
+
+set -euo pipefail
+# set -x
+
+if [ -z "${1:-}" ]; then
+    echo -n "$USAGE" >&2
+    exit 1
+fi
+
+# The parent directory to store groups of dumps in.
+DUMP_DIR=$(readlink -f "$1")
+shift
+
+if [ ! -d "$DUMP_DIR" ]; then
+    echo "DUMP_DIR '$DUMP_DIR' does not exist" >&2
+    exit 1
+fi
+
+# Ensure we're running in the directory of this script.
+SCRIPT_PATH=$(dirname "$0")
+cd "$SCRIPT_PATH"
+SCRIPT_PATH=$(pwd)
+
+# Only load library after changing to script directory.
+source lib.sh
+
+if [ -z "${LANGUAGES:-}" ]; then
+    # Load languages from config.
+    LANGUAGES=$(jq -r '(.sections_to_remove | keys | .[])' article_processing_config.json)
+fi
+# shellcheck disable=SC2086 # LANGUAGES is intentionally expanded.
+log "Selected languages:" $LANGUAGES
+
+log "Fetching run index"
+# The date of the latest dump, YYYYMMDD.
+LATEST_DUMP=$(wget 'https://dumps.wikimedia.org/other/enterprise_html/runs/' --no-verbose -O - \
+            | grep -Po '(?<=href=")[^"]*' | grep -P '\d{8}' | sort -r | head -n1)
+LATEST_DUMP="${LATEST_DUMP%/}"
+
+log "Checking latest dump $LATEST_DUMP"
+
+URLS=
+MISSING_DUMPS=0
+for lang in $LANGUAGES; do
+    url="https://dumps.wikimedia.org/other/enterprise_html/runs/${LATEST_DUMP}/${lang}wiki-NS0-${LATEST_DUMP}-ENTERPRISE-HTML.json.tar.gz"
+    if ! wget --no-verbose --method=HEAD "$url"; then
+        MISSING_DUMPS=$(( MISSING_DUMPS + 1 ))
+        log "Dump for '$lang' does not exist at '$url'"
+        continue
+    fi
+    URLS="$URLS $url"
+done
+
+if [ -z "$URLS" ]; then
+    log "No dumps available"
+    exit 16
+fi
+
+# The subdir to store the latest dump in.
+DOWNLOAD_DIR="$DUMP_DIR/$LATEST_DUMP"
+if [ ! -e "$DOWNLOAD_DIR" ]; then
+    mkdir "$DOWNLOAD_DIR"
+fi
+
+log "Downloading available dumps"
+# shellcheck disable=SC2086 # URLS should be expanded on spaces.
+wget --directory-prefix "$DOWNLOAD_DIR" --continue $URLS
+
+if [ $MISSING_DUMPS -gt 0 ]; then
+    log "$MISSING_DUMPS dumps not available yet"
+    exit 16
+fi
+
+log "Linking 'latest' to '$LATEST_DUMP'"
+LATEST_LINK="$DUMP_DIR/latest"
+ln -sf "$LATEST_DUMP" "$LATEST_LINK"
+
+# TODO: Remove old dumps?
```

Comment:
Do you want the script to handle this?

If it will be running on a cron job, then it might be good to keep 2 copies around.
Otherwise the script could delete the last dump as wikiparser is using it?

Reply:
1. Aren't files that were open before their deletion on Linux still accessible?
2. Dumps are produced regularly, right? We can set a specific schedule.
3. Script may have an option to automatically delete older dumps.

Reply:
> 1. Aren't files that were open before their deletion on Linux still accessible?

You're right, as long as `run.sh` is started before `download.sh` deletes them, it will be able to access the files.

> 2. Dumps are produced regularly, right? We can set a specific schedule.

Yes, they're started on the 1st and the 20th of each month, and finished within 3 days it looks like.

> 3. Script may have an option to automatically delete older dumps.

:+1:

Reply:
I've added a new option:

```
-D      Delete all old dump subdirectories if the latest is downloaded
```